harness/determined/launch/autodeepspeed.py

```diff
@@ -0,0 +1,242 @@
+) -> str:
+    trial_runner_hosts = ip_addresses.copy()
+    if len(ip_addresses) == 1:
+        trial_runner_hosts[0] = "localhost"
+
+    os.makedirs(hostfile_path.parent, exist_ok=True)
+    with open(hostfile_path, "w") as hostfile:
+        lines = [f"{host} slots={num_proc_per_machine}\n" for host in trial_runner_hosts]
+        hostfile.writelines(lines)
+    return trial_runner_hosts[0]
+
+
+def create_deepspeed_env_file() -> None:
+    """Create an env var export file to pass Determined vars to the deepspeed launcher.
+
+    By default, the deepspeed launcher only keeps env vars that start with one of the following
+    ["NCCL", "PYTHON", "MV2", "UCX"].
+
+    There are certain variables that we need to be set that we can pass to deepspeed using
+    a custom env vars file.
+    """
+    INCLUDE = ["PATH", "USE_DEEPSPEED", "DET_CHIEF_IP"]
+    with open(DEEPSPEED_ENVIRONMENT_NAME, "w") as f:
+        environ = cast(Dict, os.environ.copy())
+        for k, v in environ.items():
+            if k in INCLUDE:
+                f.write(f"{k}={v}\n")
+
+
+def create_run_command(master_address: str, hostfile_path: str) -> List[str]:
+    # Construct the deepspeed command.
+    deepspeed_process_cmd = [
+        "deepspeed",
+        "-H",
+        hostfile_path,
+        "--master_addr",
+        master_address,
+        "--no_python",
+        "--no_local_rank",
+        "--",
+    ]
+    return deepspeed_process_cmd
+
+
+def main(train_entrypoint: str) -> int:
+    info = det.get_cluster_info()
+    assert info is not None, "must be run on-cluster"
+    assert info.task_type == "TRIAL", f'must be run with task_type="TRIAL", not "{info.task_type}"'
+
+    # Hack: get the container id from the environment.
+    container_id = os.environ.get("DET_CONTAINER_ID")
+    assert container_id is not None, "Unable to run with DET_CONTAINER_ID unset"
+
+    # TODO: refactor websocket, data_layer, and profiling to to not use the cli_cert.
+    cert = certs.default_load(info.master_url)
+    certs.cli_cert = cert
+
+    # The launch layer should provide the chief_ip to the training code, so that the training code
+    # can function with a different launch layer in a different environment.  Inside Determined, the
+    # easiest way to get the chief_ip is with container_addrs.
+    chief_ip = info.container_addrs[0]
+
+    # Chief IP is set as an environment variable to support nested launch layers
+    os.environ["DET_CHIEF_IP"] = chief_ip
+
+    # All ranks will need to run sshd.
+    run_sshd_command = [
+        "/usr/sbin/sshd",
+        "-p",
+        str(constants.DTRAIN_SSH_PORT),
+        "-f",
+        "/run/determined/ssh/sshd_config",
+        "-D",
+    ]
+
+    if info.container_rank > 0:
+        # Non-chief machines just run sshd.
+
+        # Mark sshd containers as daemon containers that the master should kill when all non-daemon
+        # containers (deepspeed launcher, in this case) have exited.
+        api.post(
```

Comment:
@rb-determined-ai should we handle dangling sshd on chief by posting the chief container to master as well?